d0XXXr0_Halo.md

```diff
@@ -124,7 +124,10 @@ generator promise_type::get_return_object() {
 
 `begin()` must be inlinable
 ```c++
-iterator begin() { return {h}; }
+iterator begin() {
+  if (h) h.resume();
```

Comment:
This is neat. It will make an empty coroutine to be an empty range.

Reply:
as opposed to making (h != nullptr) to be a precondition for begin()

Reply:
Yes, this is needed to allow `generator<T>` default constructor to construct to an empty sequence (and also a moved-from sequence).
This is the approach I have taken with `cppcoro::generator`.

Even if you were to disallow calling `begin()` on a default-constructed sequence, you'd still need to call `h.resume()` here to allow the initially-suspended coroutine to reach either the first `co_yield` or `co_return` statement. I'd expect most compilers should be able to eliminate the `h != nullptr` check for coroutines they are directly calling/constructing anyway.